File: src/ClaudeToolbar.Core/Usage/UsageResponseParser.cs
using System.Globalization;
using System.Text.Json;

namespace ClaudeToolbar.Core.Usage;

public static class UsageResponseParser
{
    public static UsageResult Parse(string json, DateTimeOffset fetchedAt)
    {
        if (string.IsNullOrWhiteSpace(json))
            return new UsageResult.Failed("Empty response");

        JsonDocument doc;
        try
        {
            doc = JsonDocument.Parse(json);
        }
        catch (JsonException ex)
        {
            return new UsageResult.Failed($"Invalid JSON: {ex.Message}");
        }

        using (doc)
        {
            var root = doc.RootElement;
            if (root.ValueKind != JsonValueKind.Object)
                return new UsageResult.Failed("Response is not a JSON object");

            var snapshot = new UsageSnapshot(
                ReadWindow(root, "five_hour"),
                ReadWindow(root, "seven_day"),
                ReadWindow(root, "seven_day_opus"),
                ReadWindow(root, "seven_day_sonnet"),
                fetchedAt);
            return new UsageResult.Ok(snapshot);
        }
    }

    private static UsageWindow? ReadWindow(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var el) || el.ValueKind != JsonValueKind.Object)
            return null;
        if (!el.TryGetProperty("utilization", out var u) || u.ValueKind != JsonValueKind.Number)
            return null;

        DateTimeOffset? resetsAt = null;
        if (el.TryGetProperty("resets_at", out var r) && r.ValueKind == JsonValueKind.String &&
            DateTimeOffset.TryParse(r.GetString(), CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out var parsed))
        {
            resetsAt = parsed;
        }

        return new UsageWindow(Math.Clamp(u.GetDouble(), 0, 100), resetsAt);
    }
}

File: mac/Sources/ClaudeToolbarCore/Usage/UsageResponseParser.swift
import Foundation

public enum UsageResponseParser {
    public static func parse(_ json: String, fetchedAt: Date) -> UsageResult {
        let trimmed = json.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return .failed("Empty response") }

        let object: Any
        do {
            object = try JSONSerialization.jsonObject(with: Data(trimmed.utf8), options: [.fragmentsAllowed])
        } catch {
            return .failed("Invalid JSON: \(error.localizedDescription)")
        }

        guard let root = object as? [String: Any] else {
            return .failed("Response is not a JSON object")
        }

        return .ok(UsageSnapshot(
            fiveHour: readWindow(root, "five_hour"),
            sevenDay: readWindow(root, "seven_day"),
            sevenDayOpus: readWindow(root, "seven_day_opus"),
            sevenDaySonnet: readWindow(root, "seven_day_sonnet"),
            sevenDayFable: readWindow(root, "seven_day_fable") ?? readScopedWeekly(root, model: "Fable"),
            fetchedAt: fetchedAt))
    }

    private static func readWindow(_ root: [String: Any], _ name: String) -> UsageWindow? {
        guard let element = root[name] as? [String: Any] else { return nil }
        guard let raw = JSONNumber.double(element["utilization"]) else { return nil }
        var resetsAt: Date?
        if let text = element["resets_at"] as? String {
            resetsAt = ISO8601.parse(text)
        }
        return UsageWindow(utilization: min(max(raw, 0), 100), resetsAt: resetsAt)
    }

    /// Newer models have no `seven_day_<model>` field: their weekly limit is a `weekly_scoped` entry in `limits`,
    /// named by `scope.model.display_name`.
    private static func readScopedWeekly(_ root: [String: Any], model: String) -> UsageWindow? {
        guard let limits = root["limits"] as? [[String: Any]] else { return nil }
        for limit in limits where limit["kind"] as? String == "weekly_scoped" {
            let scope = limit["scope"] as? [String: Any]
            let scopeModel = scope?["model"] as? [String: Any]
            guard let name = scopeModel?["display_name"] as? String,
                  name.caseInsensitiveCompare(model) == .orderedSame,
                  let raw = JSONNumber.double(limit["percent"]) else { continue }
            let resetsAt = (limit["resets_at"] as? String).flatMap(ISO8601.parse)
            return UsageWindow(utilization: min(max(raw, 0), 100), resetsAt: resetsAt)
        }
        return nil
    }
}
